lib/auth/scopes/access/service.go

```diff
@@ -153,100 +174,278 @@ func (s *Server) CreateScopedRoleAssignment(ctx context.Context, req *scopedacce
+	// iteration, but we don't currently support resource attributes in scoped role where clauses, so
+	// can be pre-built once at the beginning of the call.
+	ruleCtx := authzContext.RuleContext()
+
+	// do a pre-check to weed out requests that definitely won't be authorized.
+	if err := authzContext.CheckerContext.CheckMaybeHasAccessToRules(&ruleCtx, scopedaccess.KindScopedRoleAssignment, types.VerbReadNoSecrets, types.VerbList); err != nil {
+		return nil, trace.Wrap(err)
 	}
 
-	return s.cfg.Reader.ListScopedRoleAssignments(ctx, req)
+	// list scoped role assignments with a filter that only passes assignments the user has access to.
+	rsp, err := s.cfg.Reader.ListScopedRoleAssignmentsWithFilter(ctx, req, func(assignment *scopedaccessv1.ScopedRoleAssignment) bool {
+		err := authzContext.CheckerContext.Decision(ctx, assignment.GetScope(), func(checker *services.SplitAccessChecker) error {
+			return checker.Common().CheckAccessToRules(&ruleCtx, scopedaccess.KindScopedRoleAssignment, types.VerbReadNoSecrets, types.VerbList)
+		})
+		return err == nil
+	})
+	return rsp, trace.Wrap(err)
 }
 
 // ListScopedRoles implements [scopedaccessv1.ScopedRoleServiceServer].
 func (s *Server) ListScopedRoles(ctx context.Context, req *scopedaccessv1.ListScopedRolesRequest) (*scopedaccessv1.ListScopedRolesResponse, error) {
-	authzContext, err := s.cfg.Authorizer.Authorize(ctx)
+	authzContext, err := s.cfg.ScopedAuthorizer.AuthorizeScoped(ctx)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	if !authz.HasBuiltinRole(*authzContext, string(types.RoleAdmin)) {
-		s.cfg.Logger.WarnContext(ctx, "user does not have permission to list scoped roles", "user", authzContext.User.GetName())
-		return nil, trace.AccessDenied("user %q does not have permission to list scoped roles", authzContext.User.GetName())
+	// build rule context for where-clause evaluation once before iteration. a resource-dependent
+	// rule context would need to be built once for each invocation of the filter function, but
+	// we don't currently support resource attributes in scoped role where clauses, so can be pre-built
+	// once at the beginning of the call.
+	ruleCtx := authzContext.RuleContext()
+
+	// do a pre-check to weed out requests that definitely won't be authorized.
+	if err := authzContext.CheckerContext.CheckMaybeHasAccessToRules(&ruleCtx, scopedaccess.KindScopedRole, types.VerbReadNoSecrets, types.VerbList); err != nil {
+		return nil, trace.Wrap(err)
 	}
 
-	return s.cfg.Reader.ListScopedRoles(ctx, req)
+	// list scoped roles with a filter that only passes roles the user has access to.
+	rsp, err := s.cfg.Reader.ListScopedRolesWithFilter(ctx, req, func(role *scopedaccessv1.ScopedRole) bool {
+		err := authzContext.CheckerContext.Decision(ctx, role.GetScope(), func(checker *services.SplitAccessChecker) error {
+			return checker.Common().CheckAccessToRules(&ruleCtx, scopedaccess.KindScopedRole, types.VerbReadNoSecrets, types.VerbList)
+		})
+		return err == nil
```

Comment:
nit: it's currently the same for all the RBAC-filtered listings, but we should really consider adding a way to check for access that doesn't signal failure with an error (that inevitably ends up calculating a stack trace) when iterating over a collection for which we might legitimately have only a handful of passes and a lot of denies.